### DISCUSSION TITLE
tutorials/best_practices/scenes_versus_scripts

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#113] [User-contributed notes]
# tutorials/best_practices/scenes_versus_scripts

We've already covered how scenes and scripts are different. Scripts define an engine class extension with imperative code, scenes with declarative code. Each system's capabilities are different as ...

https://docs.godotengine.org/en/stable/tutorials/best_practices/scenes_versus_scripts.html

**Comment by vaunakiller** [comment]
As a developer having almost little to no C++ experience I struggle to understand why [Anonymous Types](https://docs.godotengine.org/en/stable/tutorials/best_practices/scenes_versus_scripts.html#anonymous-types) section is called as such.
ChatGPT suggest the following explanation: _Creating a scene directly in script is like using anonymous types in programming: both involve dynamically defining structures without formal pre-defined templates or class names_
Is this what is implied here?

thanks <3

**Reply by Arukorstza** [reply]
It's Anonymous because it isn't defined or tied to an object/node, I think.

I honestly wouldn't worry about it.

**Reply by LiamFricker** [reply]
The main purpose in the way I interpreted it is to differentiate from the "named types".
The named types basically lets you definite a node which you can then use in future scenes while the anonymous types only create those scenes at runtime. 
Not a very good example, but I have a Custom Type made for my enemies which I can select in the editor as a baseline template to work off on..
However, if I'd like to spawn an enemy by creating a new enemy and instantiating it through a script, I would be making an Anonymous Type since the new enemy won't be saved anywhere (unless I specifically choose to have it).

**Reply by cedricao** [reply]
In my opinion, anonymous types are something that is typically lower level, associated to a programming language itself, and not a tool/engine.

Perhaps the "Defining scenes programmatically" would better fit this concept to avoid confusing readers.

For example, anonymous types in C# can be represented as follows:

```csharp
// Named type (strong typing)

class MyObj
{
    public string? Message { get; set; }
}

//...

// Using the named type
MyObj myObj = new MyObj
{
   Message = "Hello, World!",
   Description = "Hello!" // Compiler error as MyObj does not define a property 'Description'
};

// Using an anonymous type
var myObj = new 
{
   Message = "Hello, World!",
   Description = "Hello!" // OK
};
```

**Comment by pochoco24** [comment]
I'm worried about the performance difference when using statements like `Node.new()` or `add_child()` in my Classes:

"Script code like this is much slower than engine-side C++ code. Each instruction makes a call to the scripting API which leads to many "lookups" on the back-end to find the logic to execute."

Isn't this how **ALL** Godot Scripts work? or is it only slower when using methods to add or edit instances?

I hope that the section is just recommending to use Scenes to create predefined node trees over manually creating nodes via Script.

**Reply by mason-io** [reply]
I found that I was constantly worrying about performance until I started measuring things. How are you measuring your performance? What's your minimum spec target hardware and target frame time for that hardware? When you measure the performance difference between using scenes or using those statements, does it make a significant difference in your game?

**Reply by dtroitskiy** [reply]
I think the performance difference is noticeable in comparison of these two approaches, but in the actual development it can be neglected, I believe, until maybe you create hundreds of nodes every frame or something like that.
And yeah, of course using `Node.new()` and `add_child()` should be considered completely appropriate in any generic game project developement scenario.

**Comment by sauravsaurav** [comment]
=== SUMMARY =====
* Scripts : Defines Behavior 
* Scenes : Defines Structure

=> Creating objects : 
  a) Scripts : var my_node = MyNode.new()
  b) Scenes : var my_scene = MyScene.instantiate()

=> Performance : Scripts creates nodes instructions by instructions , which is slower. Scenes are faster

**Comment by Nodragem** [comment]
If I understand correctly this page is about when using "Custom Node" vs using "PackedScene". Then the Named Type section is out-topic, furthermore, I can't see a use case where add_custom_type is better than using class_name?
